mirrord/agent/src/incoming/connection/tcp_task.rs

```diff
@@ -0,0 +1,166 @@
+use std::borrow::Cow;
+
+use bytes::BytesMut;
+use mirrord_tls_util::MaybeTls;
+use tokio::{
+    io::{AsyncReadExt, AsyncWriteExt},
+    net::TcpStream,
+    sync::mpsc,
+};
+
+use crate::incoming::{
+    connection::{ConnectionInfo, IncomingIO, IncomingStreamItem},
+    error::ConnError,
+};
+
+/// Background task responsible for handling IO on redirected TCP connections.
+pub struct TcpTask {
+    pub incoming_io: Box<dyn IncomingIO>,
+    pub destination: Destination,
+}
+
+impl TcpTask {
+    /// Runs this task until the connection is closed.
+    ///
+    /// This method must ensure that the final [`IncomingStreamItem::Finished`] is always sent to
+    /// the client.
+    pub async fn run(mut self) {
+        let result = self.run_inner().await;
+
+        if let Destination::StealingClient { data_tx, .. } = &self.destination {
+            let _ = data_tx.send(IncomingStreamItem::Finished(result)).await;
+        }
+    }
+
+    async fn run_inner(&mut self) -> Result<(), ConnError> {
+        let mut source_writes = true;
+        let mut destination_writes = true;
+        let mut read_buf = BytesMut::with_capacity(64 * 1024);
```

Comment:
The `self.inconming_io` implements aysnc read and write and uses this buffer when reading. I wonder if we can do the same for `Destination`, i.e. getting rid of the `BytesMut` inside `Destination`, so we can turn the while loop below into using tokio copy io bidirectional.